### PR TITLE
Print latest block number and improve error handling

### LINKED ISSUE #757: Add printing used block id in fork when `latest` to allow reproducing

### PR DESCRIPTION
**Pull request by piotmag769**
Closes #757 

## Introduced changes

- added printing block number when `BlockId::Tag(Latest)` is specified
- improved error handling
- added `unwrap` in places where the error will be only our fault

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: crates/forge/src/running.rs [hunk 1]
```diff
@@ -229,35 +234,42 @@ pub(crate) fn run_test_case(
...
+        Err(RunnerError::CairoRunError(error)) => Ok(TestCaseSummary::Failed {
+            name: case.name.clone(),
+            msg: Some(format!(
+                "\n    {}\n",
+                error.to_string().replace(" Custom Hint Error: ", "\n    ")
+            )),
+            arguments: args,
+            fuzzing_statistic: None,
+        }),
+        Err(err) => Err(err.into()),
```
Comment (piotmag769) [inline crates/forge/src/running.rs:255]: Those are other variants of [`RunnerError`](https://github.com/starkware-libs/cairo/blob/main/crates/cairo-lang-runner/src/lib.rs#L47), I don't think they should ever happen but I'm not sure. I decided to bubble an error, it kills the whole runner anyways but with a nice error message

File: crates/forge/src/running.rs [hunk 2]
```diff
@@ -229,35 +234,42 @@ pub(crate) fn run_test_case(
         &string_to_hint,
     );
 
-    Ok(runner.run_function(
+    match runner.run_function(
```
Comment (cptartur) [inline crates/forge/src/running.rs:237]: We're doing `find_function` again here even though we have `func` variable available.

File: crates/forge/src/running.rs [hunk 3]
```diff
@@ -229,35 +234,42 @@ pub(crate) fn run_test_case(
...
-    )?)
+    ) {
+        Ok(result) => Ok(TestCaseSummary::from_run_result(result, case, args)),
+        // CairoRunError comes from VirtualMachineError which may come from HintException that originates in the TestSyscallExecutionHandler
+        Err(RunnerError::CairoRunError(error)) => Ok(TestCaseSummary::Failed {
+            name: case.name.clone(),
+            msg: Some(format!(
+                "\n    {}\n",
+                error.to_string().replace(" Custom Hint Error: ", "\n    ")
+            )),
```
Comment (cptartur) [inline crates/forge/src/running.rs:251]: I know we had this code for a while now, but what in case the error is something different than `Custom Hint Error`. Shouldn't we kill the runner as well, as we do with other errors?

File: crates/forge/src/running.rs [hunk 4]
```diff
@@ -229,35 +234,42 @@ pub(crate) fn run_test_case(
...
+            arguments: args,
+            fuzzing_statistic: None,
+        }),
+        Err(err) => Err(err.into()),
+    }
 }
 
 fn extract_test_case_summary(
```
Comment (cptartur) [inline crates/forge/src/running.rs:259]: That's a weird flow to me. `TestCaseSummary` is already extracted in this function argument. We should have this handling in one place, there's no benefit to wrapping the error in two different places.

**Review by cptartur — CHANGES_REQUESTED** [review]

File: crates/forge/src/running.rs (hunk 4 above)
Comment (piotmag769) [reply crates/forge/src/running.rs:259]: The reason for that change is to be able to add  additional info to `TestCaseSummary`, e.g. number of the block used (that's why I did this PR, see description), if we got to `run_function` call . It is a compromise between doing weird `match` statements in every place `?` is used in `run_test_case` and passing additional info to `TestCrateSummary`

File: crates/forge/src/running.rs (hunk 3 above)
Comment (piotmag769) [reply crates/forge/src/running.rs:251]: Not really, for example when reading from a file and the file does not exists this logic does not apply

File: crates/forge/src/running.rs (hunk 4 above)
Comment (cptartur) [reply crates/forge/src/running.rs:259]: I don't understand why cannot the arms of match in `run_test_case` and in `extract_test_case_summary` be combined. I'd rather have a one large match handling all errors than artificially splitting it into two functions.

File: crates/forge/src/running.rs [hunk 5]
```diff
@@ -291,12 +303,15 @@ fn get_fork_state_reader(
...
+fn get_latest_block_number(
+    url: &Url,
+) -> Result<BlockId, ProviderError<JsonRpcClientError<HttpTransportError>>> {
     let client = JsonRpcClient::new(HttpTransport::new(url.clone()));
     let runtime = Runtime::new().expect("Could not instantiate Runtime");
 
     match runtime.block_on(client.get_block_with_tx_hashes(BlockId::Tag(Latest))) {
         Ok(MaybePendingBlockWithTxHashes::Block(block)) => Ok(BlockId::Number(block.block_number)),
-        _ => Err(anyhow!("Could not get the latest block number".to_string())),
+        Ok(MaybePendingBlockWithTxHashes::PendingBlock(_)) => unreachable!(),
```
Comment (Arcticae) [inline crates/forge/src/running.rs:314]: This is out of scope

File: crates/forge/src/running.rs [hunk 6]
```diff
@@ -229,35 +234,42 @@ pub(crate) fn run_test_case(
...
-    Ok(runner.run_function(
+    match runner.run_function(
         runner.find_function(case.name.as_str())?,
         &mut test_execution_syscall_handler,
         hints_dict,
         instructions,
         builtins,
-    )?)
+    ) {
+        Ok(result) => Ok(TestCaseSummary::from_run_result(result, case, args)),
```
Comment (Arcticae) [inline crates/forge/src/running.rs:244]: Why move the parsing here? It makes whole thing a lot more confusing

**Review by Arcticae — CHANGES_REQUESTED** [review]

File: crates/forge/src/running.rs (hunk 6 above)
Comment (piotmag769) [reply crates/forge/src/running.rs:244]: Will change it, but it requires introducing additional intermediate struct. I need to be able to pass block number with a `RunResult` somehow

File: crates/forge/src/running.rs (hunk 5 above)
Comment (piotmag769) [reply crates/forge/src/running.rs:314]: It is not, before we returned a general error that would kill the whole runner. Now I want to return a specific error that occurs on rpc call, so I have to add this variant to the match statement to be able to get the error arm below. It is impossible for us to ever get `PendingBlock` here.

File: crates/forge/src/running.rs (hunk 3 above)
Comment (piotmag769) [reply crates/forge/src/running.rs]: > I generally don't understand those classes of errors. What is RunnerError? What is StateError?

External libraries errors, decided to remove them and just make sure everything that fails in `run_from_test` results in failed test (does not kill the runner)

File: crates/forge/src/running.rs (hunk 4 above)
Comment (piotmag769) [reply crates/forge/src/running.rs:264]: > I don't think running.rs should know anything about TestCaseSummary (stuff like fuzzing statistics and block number) it should be handled in an upper layer (cli layer). We should differentiate between TestCaseResult and summary and don't leak this stuff here as this feeds the complexity.

@MaksymilianDemitraszek out of scope of this PR, should be done with #839

File: crates/forge/src/running.rs [hunk 7]
```diff
@@ -151,18 +151,12 @@ fn build_syscall_handler<'a>(
...
-pub(crate) enum TestCaseRunError {
-    #[error("{0}")]
-    RunnerError(#[from] RunnerError),
-    #[error("{0}")]
-    StateError(#[from] StateError),
-    #[error("{0}")]
-    Error(#[from] anyhow::Error),
+pub(crate) struct RunResultWithInfo {
+    pub(crate) run_result: Result<RunResult, RunnerError>,
+    pub(crate) block_number_of_latest: Option<BlockNumber>,
```
Comment (Arcticae) [inline crates/forge/src/running.rs:159]:
```suggestion
    pub(crate) latest_block_number: Option<BlockNumber>,
```
I know that this isn't exactly the same semantically, but wouldn't it make it more readable?

**Review by Arcticae — CHANGES_REQUESTED** [review]

File: crates/forge/src/test_case_summary.rs [hunk 8]
```diff
@@ -23,6 +24,8 @@ pub enum TestCaseSummary {
         arguments: Vec<Felt252>,
         /// Statistic for fuzzing test
         fuzzing_statistic: Option<FuzzingStatistics>,
+        /// Number of block used if BlockId::Tag(Latest) was specified
+        block_number_of_latest: Option<BlockNumber>,
```
Comment (cptartur) [inline crates/forge/src/test_case_summary.rs:28]: I think we should follow the suit of `FuzzingStatistics` and introduce a struct for a forking parameters (e.g. name it `ForkConfigUsed`) and store this details there.

Comment (cptartur) [reply crates/forge/src/test_case_summary.rs:28]: Perhaps we could always store the block number and have a separate boolean flag for if `latest` block was used. Could make the logic simpler.

Comment (piotmag769) [reply crates/forge/src/test_case_summary.rs:28]: Out of scope: #1049

File: crates/forge/src/running.rs [hunk 9]
```diff
@@ -229,46 +227,66 @@ pub(crate) fn run_test_case(
...
-        runner.find_function(case.name.as_str())?,
+    let block_number_of_latest = if let Some(ValidatedForkConfig {
+        url: _,
+        block_id: BlockId::Tag(Latest),
+    }) = &case.fork_config
+    {
+        Some(block_info.block_number)
+    } else {
+        None
+    };
```
[the remark this answers is not in the document]
Comment (piotmag769) [reply crates/forge/src/running.rs:238]: Out of scope: #1049

File: crates/forge/src/test_case_summary.rs (hunk 8 above)
Comment (piotmag769) [reply crates/forge/src/test_case_summary.rs:28]: Agree, but imo `Option` expresses it better for now

Comment (piotmag769) [reply crates/forge/src/test_case_summary.rs:28]: I would leave it for now and wrap it in a struct when additional info will be needed to not introduce unnecessary abstraction layer

**Review by Arcticae — APPROVED** [review]

**Review by cptartur — APPROVED** [review]
